File: rossoctl/ui-v2/src/services/shipwright.ts
// Copyright 2025 IBM Corp.
// Licensed under the Apache License, Version 2.0

/**
 * Shared Shipwright build service.
 *
 * This module provides types and utilities for Shipwright builds that are used
 * by both agent and tool pages. It enables code reuse for build status tracking,
 * BuildRun management, and build finalization.
 */

/**
 * Resource type for builds
 */
export type ResourceType = 'agent' | 'tool';

/**
 * Shipwright build configuration
 */
export interface ShipwrightBuildConfig {
  buildStrategy: string;
  dockerfile: string;
  buildArgs?: string[];
  buildTimeout: string;
}

/**
 * ClusterBuildStrategy information
 */
export interface ClusterBuildStrategy {
  name: string;
  description?: string;
}

/**
 * Build status condition
 */
export interface BuildStatusCondition {
  type: string;
  status: string;
  reason?: string;
  message?: string;
  lastTransitionTime?: string;
}

/**
 * Shipwright Build status
 */
export interface ShipwrightBuildStatus {
  name: string;
  namespace: string;
  registered: boolean;
  reason?: string;
  message?: string;
}

/**
 * BuildRun phase
 */
export type BuildRunPhase = 'Pending' | 'Running' | 'Succeeded' | 'Failed';

/**
 * Shipwright BuildRun status
 */
export interface ShipwrightBuildRunStatus {
  name: string;
  namespace: string;
  buildName: string;
  phase: BuildRunPhase;
  startTime?: string;
  completionTime?: string;
  outputImage?: string;
  outputDigest?: string;
  failureMessage?: string;
  conditions: BuildStatusCondition[];
}

/**
 * Resource configuration stored in Build annotations.
 * This is generic and works for both agents and tools.
 */
export interface ResourceConfigFromBuild {
  protocol: string;
  framework: string;
  createHttpRoute: boolean;
  registrySecret?: string;
  envVars?: Array<{
    name: string;
    value?: string;
    valueFrom?: {
      secretKeyRef?: { name: string; key: string };
      configMapKeyRef?: { name: string; key: string };
    };
  }>;
  servicePorts?: Array<{
    name: string;
    port: number;
    targetPort: number;
    protocol: string;
  }>;
}

/**
 * Full Shipwright Build information including resource config and BuildRun status.
 * This is a generic type that works for both agents and tools.
 */
export interface ShipwrightBuildInfo {
  // Build info
  name: string;
  namespace: string;
  resourceType: ResourceType;
  buildRegistered: boolean;
  buildReason?: string;
  buildMessage?: string;
  outputImage: string;
  strategy: string;
  gitUrl: string;
  gitRevision: string;
  contextDir: string;

  // Latest BuildRun info
  hasBuildRun: boolean;
  buildRunName?: string;
  buildRunPhase?: BuildRunPhase;
  buildRunStartTime?: string;
  buildRunCompletionTime?: string;
  buildRunOutputImage?: string;
  buildRunOutputDigest?: string;
  buildRunFailureMessage?: string;

  // Resource configuration from annotations (generic)
  resourceConfig?: ResourceConfigFromBuild;
}

/**
 * Agent-specific build info response (for backwards compatibility).
 * Uses agentConfig instead of resourceConfig.
 */
export interface AgentShipwrightBuildInfo {
  // Build info
  name: string;
  namespace: string;
  buildRegistered: boolean;
  buildReason?: string;
  buildMessage?: string;
  outputImage: string;
  strategy: string;
  gitUrl: string;
  gitRevision: string;
  contextDir: string;

  // Latest BuildRun info
  hasBuildRun: boolean;
  buildRunName?: string;
  buildRunPhase?: BuildRunPhase;
  buildRunStartTime?: string;
  buildRunCompletionTime?: string;
  buildRunOutputImage?: string;
  buildRunOutputDigest?: string;
  buildRunFailureMessage?: string;

  // Agent configuration from annotations (agent-specific for backwards compatibility)
  agentConfig?: ResourceConfigFromBuild;
}

/**
 * Trigger BuildRun response
 */
export interface TriggerBuildRunResponse {
  success: boolean;
  buildRunName: string;
  namespace: string;
  buildName: string;
  message: string;
}

/**
 * Finalize build response
 */
export interface FinalizeBuildResponse {
  success: boolean;
  name: string;
  namespace: string;
  message: string;
}

/**
 * Finalize build request data
 */
export interface FinalizeBuildRequest {
  protocol?: string;
  framework?: string;
  envVars?: Array<{
    name: string;
    value?: string;
    valueFrom?: {
      secretKeyRef?: { name: string; key: string };
      configMapKeyRef?: { name: string; key: string };
    };
  }>;
  servicePorts?: Array<{
    name: string;
    port: number;
    targetPort: number;
    protocol: string;
  }>;
  createHttpRoute?: boolean;
  imagePullSecret?: string;
}

/**
 * Helper function to determine if a build is in progress
 */
export function isBuildInProgress(phase?: BuildRunPhase): boolean {
  return phase === 'Pending' || phase === 'Running';
}

/**
 * Helper function to determine if a build has completed (success or failure)
 */
export function isBuildComplete(phase?: BuildRunPhase): boolean {
  return phase === 'Succeeded' || phase === 'Failed';
}

/**
 * Helper function to determine if a build succeeded
 */
export function isBuildSucceeded(phase?: BuildRunPhase): boolean {
  return phase === 'Succeeded';
}

/**
 * Helper function to determine if a build failed
 */
export function isBuildFailed(phase?: BuildRunPhase): boolean {
  return phase === 'Failed';
}

/**
 * Helper function to get a human-readable build status message
 */
export function getBuildStatusMessage(buildInfo: ShipwrightBuildInfo | AgentShipwrightBuildInfo): string {
  if (!buildInfo.buildRegistered) {
    return buildInfo.buildMessage || 'Build not registered';
  }

  if (!buildInfo.hasBuildRun) {
    return 'No build run started';
  }

  switch (buildInfo.buildRunPhase) {
    case 'Pending':
      return 'Build pending...';
    case 'Running':
      return 'Build in progress...';
    case 'Succeeded':
      return 'Build completed successfully';
    case 'Failed':
      return buildInfo.buildRunFailureMessage || 'Build failed';
    default:
      return 'Unknown build status';
  }
}

/**
 * Helper function to get a status color/variant for PatternFly components
 */
export function getBuildStatusVariant(
  phase?: BuildRunPhase
): 'success' | 'warning' | 'danger' | 'info' | 'default' {
  switch (phase) {
    case 'Succeeded':
      return 'success';
    case 'Failed':
      return 'danger';
    case 'Running':
      return 'info';
    case 'Pending':
      return 'warning';
    default:
      return 'default';
  }
}

/**
 * Convert AgentShipwrightBuildInfo to generic ShipwrightBuildInfo
 */
export function toGenericBuildInfo(agentBuildInfo: AgentShipwrightBuildInfo): ShipwrightBuildInfo {
  return {
    ...agentBuildInfo,
    resourceType: 'agent',
    resourceConfig: agentBuildInfo.agentConfig,
  };
}

/**
 * Convert generic ShipwrightBuildInfo to AgentShipwrightBuildInfo (for backwards compatibility)
 */
export function toAgentBuildInfo(buildInfo: ShipwrightBuildInfo): AgentShipwrightBuildInfo {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { resourceType, resourceConfig, ...rest } = buildInfo;
  return {
    ...rest,
    agentConfig: resourceConfig,
  };
}

File: rossoctl/backend/app/models/shipwright.py
# Copyright 2025 IBM Corp.
# Licensed under the Apache License, Version 2.0

"""
Shared Pydantic models for Shipwright builds.

These models are used by both agent and tool routers for Shipwright build operations.
"""

from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel

from app.core.constants import (
    SHIPWRIGHT_DEFAULT_DOCKERFILE,
    SHIPWRIGHT_DEFAULT_TIMEOUT,
)


class ResourceType(str, Enum):
    """Type of resource being built."""

    AGENT = "agent"
    TOOL = "tool"


class ShipwrightBuildConfig(BaseModel):
    """Configuration for Shipwright builds (shared by agents and tools)."""

    # buildStrategy defaults to None to allow automatic selection based on registry type
    # (internal registries use insecure push, external use secure)
    buildStrategy: Optional[str] = None
    dockerfile: str = SHIPWRIGHT_DEFAULT_DOCKERFILE
    buildArgs: Optional[List[str]] = None  # KEY=VALUE format
    buildTimeout: str = SHIPWRIGHT_DEFAULT_TIMEOUT


class BuildSourceConfig(BaseModel):
    """Git source configuration for builds."""

    gitUrl: str
    gitRevision: str = "main"
    contextDir: str = "."
    gitSecretName: Optional[str] = None


class BuildOutputConfig(BaseModel):
    """Output image configuration for builds."""

    registry: str
    imageName: str
    imageTag: str = "latest"
    pushSecretName: Optional[str] = None


class BuildStatusCondition(BaseModel):
    """Build status condition."""

    type: str
    status: str
    reason: Optional[str] = None
    message: Optional[str] = None
    lastTransitionTime: Optional[str] = None


class ClusterBuildStrategyInfo(BaseModel):
    """Information about a ClusterBuildStrategy."""

    name: str
    description: Optional[str] = None


class ClusterBuildStrategiesResponse(BaseModel):
    """Response containing available ClusterBuildStrategies."""

    strategies: List[ClusterBuildStrategyInfo]


class ShipwrightBuildListItem(BaseModel):
    """Summary of one Shipwright Build CR (Rossoctl agent or tool pipeline)."""

    name: str
    namespace: str
    resourceType: str = ""  # "agent" or "tool" from rossoctl.io/type label
    registered: bool = False
    strategy: str = ""
    gitUrl: str = ""
    gitRevision: str = ""
    contextDir: str = ""
    outputImage: str = ""
    creationTimestamp: Optional[str] = None


class ShipwrightBuildListResponse(BaseModel):
    """List of Rossoctl-associated Shipwright Build resources."""

    items: List[ShipwrightBuildListItem]


class ShipwrightBuildStatusResponse(BaseModel):
    """Response containing Shipwright Build status information."""

    name: str
    namespace: str
    registered: bool
    reason: Optional[str] = None
    message: Optional[str] = None


class ShipwrightBuildRunStatusResponse(BaseModel):
    """Response containing Shipwright BuildRun status information."""

    name: str
    namespace: str
    buildName: str
    phase: str  # Pending, Running, Succeeded, Failed
    startTime: Optional[str] = None
    completionTime: Optional[str] = None
    outputImage: Optional[str] = None
    outputDigest: Optional[str] = None
    failureMessage: Optional[str] = None
    conditions: List[BuildStatusCondition]


class ResourceConfigFromBuild(BaseModel):
    """Resource configuration stored in Build annotations (generic for agent/tool)."""

    protocol: str = "a2a"
    framework: str = "LangGraph"
    createHttpRoute: bool = False
    registrySecret: Optional[str] = None
    envVars: Optional[List[Dict[str, Any]]] = None
    skills: Optional[List[str]] = None
    servicePorts: Optional[List[Dict[str, Any]]] = None
    # Per-workload resource overrides stashed at form-submit time. Declared here
    # because this model drops undeclared annotation keys (Pydantic's default
    # extra="ignore"), so a field absent from the model is invisible to callers
    # that read it back via model_dump() during Shipwright finalize.
    k8sResourceLimits: Optional[Dict[str, str]] = None
    k8sResourceRequests: Optional[Dict[str, str]] = None
    # AuthBridge layer-3 plugin composition stashed at form-submit time. Declared
    # here for the same reason as k8sResourceLimits above — undeclared annotation
    # keys are dropped by this model's default extra="ignore".
    pluginPreset: Optional[str] = None
    plugins: Optional[List[str]] = None
    onError: Optional[str] = None


class ShipwrightBuildInfoResponse(BaseModel):
    """Full Shipwright Build information including resource config and latest BuildRun status.

    This model is shared between agents and tools. The resourceType field indicates
    which type of resource this build is for.
    """

    # Build info
    name: str
    namespace: str
    resourceType: str  # "agent" or "tool"
    buildRegistered: bool
    buildReason: Optional[str] = None
    buildMessage: Optional[str] = None
    outputImage: str
    strategy: str
    gitUrl: str
    gitRevision: str
    contextDir: str

    # Latest BuildRun info (if any)
    hasBuildRun: bool = False
    buildRunName: Optional[str] = None
    buildRunPhase: Optional[str] = None  # Pending, Running, Succeeded, Failed
    buildRunStartTime: Optional[str] = None
    buildRunCompletionTime: Optional[str] = None
    buildRunOutputImage: Optional[str] = None
    buildRunOutputDigest: Optional[str] = None
    buildRunFailureMessage: Optional[str] = None

    # Resource configuration from annotations (generic dict for flexibility)
    resourceConfig: Optional[Dict[str, Any]] = None
